### PR TITLE
Add Interpreter parameter to Manager and Repeater class

### PR DESCRIPTION
**Pull request by carlosmhd27**
This pull request adds support for specifying an interpreter (such as Python, Julia, or R) when running executables in both the `Repeater` and `Manager` classes. This feature allows users to run scripts that require an interpreter, which is important for Windows users as they need to specify an interpreter to run PowerShell or bash commands, with flexible specification as either a string or list of arguments. The documentation and tests are updated to cover this new functionality, and internal code is refactored for clarity and maintainability.

**Interpreter support for executables:**

* Added an `interpreter` parameter to both `Repeater` and `Manager`, allowing executables to be run through an interpreter (e.g., `python`, `conda run ...`). This can be specified as a string or a list of strings, and is prepended to the command used in `subprocess.run`. [[1]](diffhunk://#diff-bdcf7842ef59715a6b24f7f5cef089d67a7ad31eb98167682f1d45a058a41e24L32-R46) [[2]](diffhunk://#diff-bdcf7842ef59715a6b24f7f5cef089d67a7ad31eb98167682f1d45a058a41e24R174) [[3]](diffhunk://#diff-bdcf7842ef59715a6b24f7f5cef089d67a7ad31eb98167682f1d45a058a41e24L181-R237)
* Implemented getter and setter methods for the `interpreter` property in both classes, with validation to ensure correct types. [[1]](diffhunk://#diff-bdcf7842ef59715a6b24f7f5cef089d67a7ad31eb98167682f1d45a058a41e24R60-R63) [[2]](diffhunk://#diff-bdcf7842ef59715a6b24f7f5cef089d67a7ad31eb98167682f1d45a058a41e24R76-R86) [[3]](diffhunk://#diff-bdcf7842ef59715a6b24f7f5cef089d67a7ad31eb98167682f1d45a058a41e24R272-R276) [[4]](diffhunk://#diff-bdcf7842ef59715a6b24f7f5cef089d67a7ad31eb98167682f1d45a058a41e24R290-R300)
* Updated the logic in `run` (Repeater) and `create` (Manager) to use the interpreter when launching the executable. This includes creating private function `__execute` as a helper to execute the executable. [[1]](diffhunk://#diff-bdcf7842ef59715a6b24f7f5cef089d67a7ad31eb98167682f1d45a058a41e24L85-R109) [[2]](diffhunk://#diff-bdcf7842ef59715a6b24f7f5cef089d67a7ad31eb98167682f1d45a058a41e24R187) [[3]](diffhunk://#diff-bdcf7842ef59715a6b24f7f5cef089d67a7ad31eb98167682f1d45a058a41e24L318-R378) [[4]](diffhunk://#diff-bdcf7842ef59715a6b24f7f5cef089d67a7ad31eb98167682f1d45a058a41e24R392-R409)

**Documentation updates:**

* Expanded docstrings for `Manager` to document the new `interpreter` argument, how to use it, and provided usage examples. [[1]](diffhunk://#diff-bdcf7842ef59715a6b24f7f5cef089d67a7ad31eb98167682f1d45a058a41e24L181-R237) [[2]](diffhunk://#diff-bdcf7842ef59715a6b24f7f5cef089d67a7ad31eb98167682f1d45a058a41e24L206-R250) [[3]](diffhunk://#diff-bdcf7842ef59715a6b24f7f5cef089d67a7ad31eb98167682f1d45a058a41e24L217-R261)

**Testing improvements:**

* Added new tests to verify that both `Manager` and `Repeater` work correctly with an interpreter. [[1]](diffhunk://#diff-cde49e23f2bc70f08c83299b936d25055a9175037aa4d96378728b12cac90c3cR29-R48) [[2]](diffhunk://#diff-cde49e23f2bc70f08c83299b936d25055a9175037aa4d96378728b12cac90c3cL85-R139)
* Updated existing tests to avoid side effects and ensure proper cleanup of temporary files and directories.

**Limitations**

* Tests do not work on Windows, this needs further changes.
*  If the parameter `interpreter` is passed as a string, any whitespace will be used for splitting the string. This creates problems when trying to set the interpreter using a path containing whitespaces. This is evident on Windows, as git bash lives in `C:\Program Files\Git\usr\bin\bash.exe`.

### CODE REVIEW

**Comment by mwiesenberger** [comment]
Hi sorry I only found time now to review this. I appreciate the detailed description

**Comment by mwiesenberger** [comment]
I like the solution that you found. I would go a step further and simplify it to allow the executable parameter to be a string or list of strings (as you pointed out in the issue description). All that either Manager or Repeater class are doing is forward the `interpreter` and `executable` variables to `subprocess.run` anyway (and doing nothing else with those variables, so there is no real reason to keep them as separate variables in class.  In that spirit I also suggest to not split the given string using `string.split()`. As you point out this may have some unintended consequences on Windows  and effectively, we're just forwarding stuff so we should probably stay as close as possible to the interface of  `subprocess.run`  (which accepts string or list as first argument) Any user can just call `simplesimdb.Manager(executable = "bla bla bla".split())` themeselves if they wish so

File: simplesimdb.py [hunk 1]
```diff
@@ -29,12 +29,24 @@ class Repeater:
...
+        interpreter: str | list[str] | None = None,
     ):
         """Set the executable and files to use in the run method"""
         self.executable = executable
         self.inputfile = inputfile
         self.outputfile = outputfile
+        if interpreter is None or len(interpreter) == 0:
+            self.interpreter = []
+        elif type(interpreter) is str:
+            self.interpreter = interpreter.split()
```
Comment (mwiesenberger) [inline simplesimdb.py]: As commented on the pull-request  I would prefer just
`self.interpreter = [interpreter]`

File: simplesimdb.py [hunk 2]
```diff
@@ -29,12 +29,24 @@ class Repeater:
     """
 
     def __init__(
-        self, executable="./execute.sh", inputfile="temp.json", outputfile="temp.nc"
+        self,
+        executable="./execute.sh",
+        inputfile="temp.json",
+        outputfile="temp.nc",
+        interpreter: str | list[str] | None = None,
```
Comment (mwiesenberger) [inline simplesimdb.py]: I would prefer to just use the already existing `executable` variable instead of the new `interpreter` 
i.e.
`executable: str | list[str] |None = "./execute.sh",`

Or do you see a problem with that solution?

File: simplesimdb.py [hunk 3]
```diff
@@ -178,19 +212,38 @@ def __init__(
...
+            (optional) If your executable is not directly runnable but needs to be
+            called via an interpreter (e.g. python, julia, R, powershell, etc.) you can
+            specify the interpreter here as a whitespace separated string or list of
+            strings. For example if your executable is a python script you can set
+            interpreter="python.exe" and the code will be called as
+            subprocess.run(["python.exe", executable, ...],...)
+            You can also include interpreter arguments here. For example to run in
+            a specific conda environment, set interpreter=["conda", "run", "-n",
+            "myenv"] or interpreter="conda run -n myenv" and the code will be called
+            as subprocess.run(["conda", "run", "-n", "myenv", executable, ...],...)
```
Comment (mwiesenberger) [inline simplesimdb.py]: I like this description. If we end up using `executable` this can just be merged

File: simplesimdb.py [hunk 4]
```diff
@@ -346,6 +398,24 @@ def create(
 
             return ncfile
 
+    def __execute(self, js, ncfile, n=0, stdout="ignore"):
+        """Helper function to run the executable with the correct arguments
+
+        This is used in create to keep the code cleaner and reduce identation depth.
```
Comment (mwiesenberger) [inline simplesimdb.py:361]: Good idea

File: test_simplesimdb.py [hunk 5]
```diff
@@ -27,6 +27,30 @@ def test_creation():
     assert not os.path.isdir(m.directory)
 
 
+def test_creation_with_interpreter():
+    print("TEST CREATION WITH INTERPRETER")
+    executable = (
+        "import sys; import subprocess; "
+        "subprocess.run(['cp', sys.argv[1], sys.argv[2]])"
+    )
```
Comment (mwiesenberger) [inline test_simplesimdb.py:37]: I love how you use the interface: The "script" can now just be a python string. I can see how this opens up many new use cases. Very cool

Comment (mwiesenberger) [reply test_simplesimdb.py:37]: I think if the `interpreter` and `executable` variables are merged then one could do
```python
script = (
        "import sys; import subprocess; "
        "subprocess.run(['cp', sys.argv[1], sys.argv[2]])"
    )
executable = ["python", "-c", script]
# or
executable = [*"python -c".split(), script]
```

Comment (mwiesenberger) [reply test_simplesimdb.py:37]: which puts a tiny bit more work on the user but then we do not need to handle the edge cases

File: test_simplesimdb.py [hunk 6]
```diff
@@ -82,13 +106,49 @@ def test_named_creation():
 
 def test_repeater():
     print("TEST REPEATER")
-    m = sim.Repeater("touch", "temp.json", "temp.nc")
+    os.makedirs("temp_repeater", exist_ok=True)
+    m = sim.Repeater("touch", "temp_repeater/temp.json", "temp_repeater/temp.nc")
```
Comment (mwiesenberger) [inline test_simplesimdb.py:110]: I am guessing it is cleaner to work in a subdirectory here? Or were there problems with putting it at the working directory?

File: test_simplesimdb.py [hunk 7]
```diff
@@ -82,13 +106,49 @@ def test_named_creation():
...
+        "subprocess.run(['echo', sys.argv[1], sys.argv[2]])"
+    )
+    m.run(inputdata, error="display", stdout="display")
+    m.clean()
+    assert not os.path.isfile("temp_repeater_interpreter/temp.json")
+    assert not os.path.isfile("temp_repeater_interpreter/temp.nc")
+    temp_folder_exists = os.path.isdir("temp_repeater_interpreter")
+    temp_folder_empty = not os.listdir("temp_repeater_interpreter")
+    if temp_folder_exists and temp_folder_empty:
+        os.rmdir("temp_repeater_interpreter")
```
Comment (mwiesenberger) [inline test_simplesimdb.py:154]: Nice tests, I guess on Windows one somehow needs to recognize the operating system in code and run a different test. This can be a separate issue thogh

File: simplesimdb.py (hunk 2 above)
Comment (carlosmhd27) [reply simplesimdb.py]: I think it makes sense to keep everything in the executable, it expands possibilities and if we do not split the string, it keeps compatibility with previous versions.

File: test_simplesimdb.py (hunk 7 above)
Comment (carlosmhd27) [reply test_simplesimdb.py:154]: I will open a new issue and start working on it.

File: test_simplesimdb.py (hunk 5 above)
Comment (carlosmhd27) [reply test_simplesimdb.py:37]: I agree, I like when the script does things for you, like splitting the string, but it may cause some problems down the line, specifically with paths containing whitespaces.

Comment (mwiesenberger) [reply test_simplesimdb.py:37]: I think since we're just forwarding all things to `subprocess.run` whatever problem appears with paths containing whitespaces, is not our problem but the problem of `subprocess.run`

File: test_simplesimdb.py [hunk 8]
```diff
@@ -17,7 +17,7 @@ def test_creation():
     print("TEST CREATION")
     m = sim.Manager(directory="creation_test", executable="cp", filetype="json")
     assert m.directory == "creation_test"
-    assert m.executable == "cp"
+    assert m.executable == ["cp"]
```
Comment (mwiesenberger) [inline test_simplesimdb.py]: hmm this seems a bit unexpected and might brake code. I wonder if the cleanest way to doing it is to resolve if executable is a string or list of strings only in the `__execute` function and otherwise store executable exactly how the user is setting it

Comment (mwiesenberger) [reply test_simplesimdb.py]: Do you think you can fix this? Otherwise I can also try
